dds/DCPS/transport/framework/TransportReceiveStrategy_T.cpp

```diff
@@ -78,6 +78,172 @@ TransportReceiveStrategy<TH, DSH>::check_header(const DSH& /*header*/)
+
+  ACE_INET_Addr remote_address;
+  bool stop = false;
+  ssize_t bytes_remaining = receive_bytes(iov,
+                                          1,
+                                          remote_address,
+                                          fd,
+                                          stop);
+
+  //ACE_DEBUG((LM_DEBUG, "TransportReceiveStrategy[%@]::handle_simple_dds_input(%d) - read %d bytes (stop = %C)\n", this, fd, bytes_remaining, stop ? "true" : "false"));
+
+  cur_rb->wr_ptr(bytes_remaining);
+
+  if (stop) {
+    return 0;
+  }
+
+  if (bytes_remaining < 0) {
+    relink();
+    return -1;
+  }
+
+  if (bytes_remaining == 0) {
+    if (gracefully_disconnected_) {
+      return -1;
+    } else {
+      relink();
+      return -1;
+    }
+  }
+
+  if (!pdu_remaining_) {
+    receive_transport_header_.length_ = static_cast<ACE_UINT32>(bytes_remaining);
+  }
+
+  receive_transport_header_ = *cur_rb;
+  if (!receive_transport_header_.valid()) {
+    cur_rb->reset();
+    ACE_ERROR_RETURN((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: TransportHeader invalid.\n")), -1);
+  }
+
+
+  bytes_remaining = receive_transport_header_.length_;
+  if (!check_header(receive_transport_header_)) {
+    return 0;
+  }
+
+  while (bytes_remaining > 0) {
+    data_sample_header_.pdu_remaining(bytes_remaining);
+    data_sample_header_ = *cur_rb;
+    bytes_remaining -= data_sample_header_.marshaled_size();
+    if (!check_header(data_sample_header_)) {
+      return 0;
+    }
+    size_t dsh_ml = data_sample_header_.message_length();
+    ACE_Message_Block* current_sample_block = 0;
+    ACE_NEW_MALLOC_RETURN(
+      current_sample_block,
+      (ACE_Message_Block*) mb_allocator_.malloc(sizeof(ACE_Message_Block)),
+      ACE_Message_Block(
+        cur_rb->data_block()->duplicate(),
+        0,
+        &mb_allocator_),
+      -1);
+    current_sample_block->rd_ptr(cur_rb->rd_ptr());
+    current_sample_block->wr_ptr(current_sample_block->rd_ptr() + dsh_ml);
+    cur_rb->rd_ptr(current_sample_block->rd_ptr() + dsh_ml);
+    bytes_remaining -= dsh_ml;
+    ReceivedDataSample rds(current_sample_block);
+    if (data_sample_header_.into_received_data_sample(rds)) {
+
+      if (data_sample_header_.more_fragments() || receive_transport_header_.last_fragment()) {
+        VDBG((LM_DEBUG,"(%P|%t) DBG:   Attempt reassembly of fragments\n"));
+
+        if (reassemble(rds)) {
+          VDBG((LM_DEBUG,"(%P|%t) DBG:   Reassembled complete message\n"));
+          deliver_sample(rds, remote_address);
+        }
+        // If reassemble() returned false, it takes ownership of the data
+        // just like deliver_sample() does.
+
+      } else {
+        deliver_sample(rds, remote_address);
+      }
+    }
+
+    // For the reassembly algorithm, the 'last_fragment_' header bit only
+    // applies to the first DataSampleHeader in the TransportHeader
+    receive_transport_header_.last_fragment(false);
+  }
+
+
+  if (cur_rb->data_block()->reference_count() > 1) {
```

Comment:
Are there potential data races here?

Reply:
I don't think so, as long as deliver_sample() is adding a reference before returning. There's a chance we'd inadvertently free / reallocate when not strictly necessary, but that should still be safe.